### PR TITLE
WT-9805 Move deleting the history store record to __wt_rec_upd_select

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

File: src/reconcile/rec_visibility.c [hunk 1]
```diff
@@ -182,6 +182,134 @@ __rec_append_orig_value(
...
+#endif
+
+    if (r->hs_cursor == NULL)
+        WT_RET(__wt_curhs_open(session, NULL, &r->hs_cursor));
+    hs_read_committed = F_ISSET(r->hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
+    F_SET(r->hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
+
+    /* No need to delete from the history store if it is already obsolete. */
+    if (delete_tombstone != NULL && __wt_txn_upd_visible_all(session, delete_tombstone)) {
+        ret = 0;
```
Comment (etienneptl) [inline src/reconcile/rec_visibility.c:206]: I guess it is to make it more explicit, but don't we have the guarantee that `ret` is `0` here?

File: src/reconcile/rec_visibility.c [hunk 2]
```diff
@@ -182,6 +182,134 @@ __rec_append_orig_value(
...
+                        key->data = WT_INSERT_KEY(ins);
+                        key->size = WT_INSERT_KEY_SIZE(ins);
+                    }
+                    break;
+                default:
+                    WT_ERR(__wt_illegal_value(session, r->page->type));
+                }
+                WT_ERR(__rec_delete_hs_record(session, r, key, delete_upd, delete_tombstone));
+                if (r->page->type == WT_PAGE_COL_FIX)
+                    printf("Remove hs record %" PRIu64 "\n", WT_INSERT_RECNO(ins));
```
Comment (etienneptl) [inline src/reconcile/rec_visibility.c:296]: Is that `printf` for debugging?

File: src/reconcile/rec_visibility.c [hunk 3]
```diff
@@ -182,6 +182,134 @@ __rec_append_orig_value(
...
+                break;
+            }
+        }
+    }
+
+    /*
+     * If we see a restored tombstone from the history store, we must also see a restored update
```
Comment (etienneptl) [inline src/reconcile/rec_visibility.c:303]: The `must` makes me wonder if we want instead `WT_ASSERT_ALWAYS`, what do you think?

Comment (quchenhao) [reply src/reconcile/rec_visibility.c]: done

File: src/reconcile/rec_visibility.c (hunk 2 above)
Comment (quchenhao) [reply src/reconcile/rec_visibility.c]: removed

File: src/reconcile/rec_visibility.c (hunk 1 above)
Comment (quchenhao) [reply src/reconcile/rec_visibility.c]: removed

**Comment by etienneptl** [comment]
Thanks for addressing my comments, I will do another round tomorrow. Just in case, the [failing test](https://spruce.mongodb.com/task/wiredtiger_windows_64_unit_test_bucket05_patch_2537dbfc7938d44113f355000c677bc8badb2969_631dc21b2fbabe310e9490e1_22_09_11_11_10_19/logs?execution=0) is a known failure.

File: src/history/hs_rec.c [hunk 4]
```diff
@@ -407,6 +347,19 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_MULTI *mult
...
+             upd = upd->next)
+            ;
+
+        /* No update to insert to history store. */
+        if (upd == NULL)
+            continue;
+
+        /* Updates have already been inserted to the history store. */
+        if (F_ISSET(upd, WT_UPDATE_HS))
+            continue;
```
Comment (agorrod) [inline src/history/hs_rec.c:361]: Why are these necessary now? It feels like this change requires an update to the header as well.

**Review by agorrod — COMMENTED** [review]
I've done a partial review that I don't want to lose, so I'm going to post it here.

After a conversation with Etienne and Chenhao - Chenhao is going to code up a different approach to the problem, and we can compare the solutions.

File: src/btree/bt_cursor.c [hunk 5]
```diff
@@ -612,6 +612,8 @@ __wt_btcur_reset(WT_CURSOR_BTREE *cbt)
     WT_STAT_CONN_DATA_INCR(session, cursor_reset);
 
     F_CLR(cursor, WT_CURSTD_KEY_SET | WT_CURSTD_VALUE_SET);
+    cbt->upd_value->type = WT_UPDATE_INVALID;
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/btree/bt_cursor.c:615]: It is just for safety and consistent values in the core dump. I found in core dump, this value is not initialized and it is confusing.

File: src/history/hs_rec.c (hunk 4 above)
Comment (quchenhao) [reply src/history/hs_rec.c:361]: This is to revert the change made in WT-9268 as we are not removing history store record here.

**Comment by quchenhao** [comment]
Close in favour of #8262
